Add series deletion with compactor proposal (#5436)
thanos-io/thanos: docs/proposals-accepted/202206-block-series-deletion-compactor.md

```go
// Result for the processed result for each tombstone, each block.
type MemTombstones struct {
    intvlGroups map[storage.SeriesRef]Intervals
    mtx         sync.RWMutex
}

// Result for the processed result for each block.
// Key: global tombstone ID, Value: processed result for each tombstone.
type BlockTombstones map[ulid.ULID]*MemTombstones
```

Since all the blocks and tombstones are immutable, this result can be easily cached in memory or in filesystem. Ideally for each tombstone, each block we just need to calculate once. Since `BlockTombstones` is a map, it is easy to add or delete one key if users want to add more tombstones or remove one tombstone.

During the query time, the cached tombstone will be used. As it knows which `SeriesRef`, which intervals to apply, it is very straightforward to mask series.

Pros:
* More efficient during query time, less processing required.
* As data can be cached there is not much computation required.

Cons:
* More memory required.
* If read is much less than write, it is inefficient as the pre-calculated result won't be used much.

> From my perspective, option 2 is better because data can be cached.

Reviewer: As I reader, I wonder what is `option 2` here. You mentioned "options" before, but you didn't number them 1 or 2. I am confused whether this is a reference to `Option 2: New API to Match Tombstones` (inside the tombstone deletion heading). 

I recommend numbering the options `Mask Series At Query Time` and `Mask Series At Tombstone Sync Time`, like done for the discussions about deletion with Compactor.

I believe that putting your recommendation under a separate heading would be great to give it evidence.

Author: Thanks! That makes total sense and I should really do it from the beginning. 💯

* Series are deleted at normal compaction. This might need to wait a long time and hard to apply for blocks at the max compaction level.
* Rewrite blocks separately. This needs more coordination with compaction to avoid duplicate/unnecessary work.

To overcome those challenges, we propose two options.

#### Option 1: Compactor Scans Blocks

Like store gateway, compactor can also scan block indexes to find the blocks matching the tombstones. Similar to the `Mask Series At Tombstone Sync Time` of store gateway, compactor can calculate the processed tombstones for each block, each tombstone and cache the results locally. As blocks and tombstones are immutable, the result needs to be calculated only once and then reused easily. As the Prometheus format tombstones are calculated, during the compaction time the compactor just needs to generate the tombstone file in the filesystem and the TSDB compactor will apply the deletion.

There are two approaches about how to scan block indexes:
1. Download all the blocks locally and check against the tombstones.
2. Use part of the store gateway code to scan the block index and chunks on the object storage to generate `MemTombstones` result.

> Approaches 2 should be better as only block indexes need to be downloaded, chunks data are reading from the object storage on the fly so disk space can be saved. As compactor is an offline component, it is acceptable to read from the object storage rather than downloading the whole data.

Reviewer: Typo here in `approaches`, should be `approach`.

}

message TombstoneRequest {
  int64 min_time = 1;
  int64 max_time = 2;
  repeated LabelMatcher matchers = 3 [(gogoproto.nullable) = false];
}

message TombstoneResponse {
  repeated SeriesIntervals series_intervals = 1;
}

message SeriesIntervals {
  int64 series_ref = 1;
  repeated Interval intervals = 2;
}

message Interval {
  int64 start = 1;
  int64 end = 2;
}
```

After getting the query result, the compactor does almost the same thing as the previous option: generate the tombstone file during compaction time.

The drawback for this approach is that we are inventing a new API for this purpose only, which might be a bit of pain in the future.

#### Option 3: Delete Series using Rewrite Logic

The previous two options do the series deletion at compaction time, this option does actual series deletion using block rewrite. For the index scanning, it is still required to choose either option 1 or 2 to avoid unncessary rewrite even if the block doesn't have any matched series.

This is not so efficient as rewrite is a different process than compaction. We need to add this step and make sure rewrite can work together with compaction and no redundant work is done.

> Overall, among the 3 options, option 1 is better as it is easier to implement and straightforward.

Reviewer: Same comment here about giving more evidence to your recommendation by putting it under a `My recommendation` (or something like this) heading. This is buried inside the description of an option and it's not part of the option itself.

The exception is for blocks that at max compaction level, no compaction will happen to it and the tombstones will remain forever if the ratio is lower than the deletion threshold. In this case the planner can check the block compaction level first, if it is max compaction level and the tombstone exists for some period of time, then do compaction for that block.

#### Tombstone File Deletion

After one compaction iteration is done, metadata will be synced again and compactor starts doing garbage collection. This time the compactor can sync tombstones again and check is there any matched block for the tombstone file. If no, then the tombstone file can be deleted.

For any new blocks that are compacted by blocks along with some tombstone requests, we want to avoid calculating tombstones on it again as those series are deleted from it already. We propose a new field in the `meta.json` file called `tombstone_applied` to indicate what tombstones have been applied to the block.

Tombstones will skip blocks that have already been applied with its ID.

### Tombstones sharding

> Unlike blocks, we cannot shard tombstones simply using its ULIDs because each tombstone is able to match all blocks from the object storage. When compactor sharding is enabled, it still needs to check all the tombstones in the shard to find the matched blocks.

Author: The tricky thing of only sharding blocks, not tombstones is that we cannot know whether all matched blocks are processed for the given tombstone as each compactor only has information about its own blocks

## Conclusions

* For store gateway, it masks series at tombstone sync time.

> * For compactor, it scans block indexes directly and caches the tombstones results. During compaction time, it generates the tombstone file and series deletion is done by the TSDB compactor.

Reviewer: I think it's valuable to link back to the headings where these options where described, so that in the future we can quickly jump to this section and then back to the options that were implemented.

---
type: proposal
title: Block Series Deletion with Compactor
status: accepted
owner: yeya24
menu: proposals-accepted
---

* **Owners:**
    * @yeya24

## Summary

This design document continues the feature of series deletion for object storage discussed in [proposal](https://thanos.io/tip/proposals-accepted/202012-deletions-object-storage.md/). This proposal is based on the previous work and focuses more on the actual implementation plan and the unaddressed questions from the original proposal like performing the actual deletion in the object storage using Thanos compactor.

## Goals

> * Make it more clear about the main workflow of the series deletion process.

Reviewer 2: nit:

```suggestion
* Make the main workflow of the series deletion process clearer.
```

* Perform the actual series deletion using compactor in a scalable way.

## Non-Goals

* Allow setting different retentions per series. This feature can be built on top of this work ideally but it is out of scope of this proposal.

## How

The main workflow of the series deletion will look like this:

> 1. User create a tombstone file via a CLI tool and the tombstone file is uploaded to the object storage.

Reviewer 3: Will the CLI do this, or should it be automated through a CI pipeline?

Author: Yes the cli will upload the tombstone file to the bucket directly. Similar to other bucket tools

2. Store Gateway syncs tombstone files periodically and makes the block series based on the tombstones.

> 3. Tombstone files have a pending period, which can be either specified by each tombstone or configured globally. Before the pending period is over, the tombstone file will remain in the object storage and user can remove a tombstone file using the CLI tool. Once the pending period is over, the compactor starts processing the tombstones and performs deletion.

Reviewer 3: Is the `pending period` similar to the consistency delay in compactor/store-gw?

> 3. Tombstone files have a pending period, which can be either specified by each tombstone or configured globally. Before the pending period is over, the tombstone file will remain in the object storage and user can remove a tombstone file using the CLI tool. Once the pending period is over, the compactor starts processing the tombstones and performs deletion.

Reviewer 2: What is the rationale for having a pending period?

Author: Users can undo the tombstone if they change their mind.

4. For each tombstone file, once all the matched blocks are processed, compactor will delete the tombstone file from the object storage.

### Tombstone File

The design of the tombstone file mainly follows the previous proposal. Tombstone file is immutable. It is using a custom JSON format, single file per request and global, which means it applies to all the blocks. For the detailed considerations and tradeoffs, please refer to the original proposal.

The file format is as follows:
``` bash
cat 01G61GC0Q957TVZ2FC8S6Y2ZK7.json
{
        "ulid": "01G61GC0Q957TVZ2FC8S6Y2ZK7",
        "matchers": [
                "__name__=\"k8s_app_metric39\"",
                "cluster=\"us\"",

>                 "__block_id=\"01G61D0Q4BHTRC7HW08YRD0KGS\""

Author: Yes. This is something I added. But block ID is an optional field, which means users can choose to not add it and then the matchers can match all blocks. If they choose to add it then they can delete the specific block they want to. (Even multiple blocks using =~)

>                 "__block_id=\"01G61D0Q4BHTRC7HW08YRD0KGS\""

Reviewer 3: Can we use something like the prometheus relabel configs, or is that too complicated for now?

>                 "__block_id=\"01G61D0Q4BHTRC7HW08YRD0KGS\""

Author: Hi Filip, if I understand this correctly, the tombstone file is global in the bucket and each tombstone file can match all the blocks (if no external label marcher and block id marchers are specified).
So if we are talking about sharding, store gateway and compactor need to gather all tombstones to match their blocks. Their blocks are sharded already, but tombstones are not.

That might be an issue when sharded compactors are running. Because each compactor only has view for its own blocks, we don't know whether a tombstone file has been applied to all blocks or not. So some kind of compactor coordinations are needed. WDYT about this issue?

>                 "__block_id=\"01G61D0Q4BHTRC7HW08YRD0KGS\""

Author: If we want to shard on tombstones then we might need per block tombstone file…

>                 "__block_id=\"01G61D0Q4BHTRC7HW08YRD0KGS\""

Reviewer 3: What I had in mind was to use relabel configs instead of explicit matchers to keep or drop individual series. But I guess that will increase the scope of the proposal to allow complete series rewriting in compactor, not just deletion. So something to keep in mind for the future, but too complicated for implementing in the first version.

>                 "__block_id=\"01G61D0Q4BHTRC7HW08YRD0KGS\""

Author: Oh I see what you meant. Sorry I was talking about different things... 
Similar idea has been mentioned in https://github.com/thanos-io/thanos/issues/4941. We supported this in `bucket rewrite tool` and I proposed similar things to Prometheus https://github.com/prometheus/prometheus/pull/9413 before.

But indeed this has a different scope. Maybe it is better to think about it when we are working on the per-series retention period feature on compactor https://github.com/thanos-io/thanos/issues/903.

>                 "__block_id=\"01G61D0Q4BHTRC7HW08YRD0KGS\""

Reviewer 4: Yes, we discussed this - we could add new field for tombstones to support rewrites at some point.

I like block ID thingy @yeya24  - we could add that. Just let's fix the sentence that it applies to all blocks, unless block ID is specified

>                 "__block_id=\"01G61D0Q4BHTRC7HW08YRD0KGS\""

Author: > ust let's fix the sentence that it applies to all blocks, unless block ID is specified

The matchers support matching external labels, too for multi-tenancy purposes. I think we can still call it global and it applies to all blocks. Just we can tell that the matchers won't match some blocks so we just filter them out. Like how we do requests filtering on store gateway.

        ],
        "minTime": -62167219200000,
        "maxTime": 253402300799000,
        "author": "Ben Ye",
        "reason": "not specified"
}
```

Similar to the block `meta.json` file, the tombstone file is named using ULID which is basically its creation time. For other fields we have:

> * **label matchers**: label matchers to match the block and series. Like this example, blocks can be matched using external labels and blocks can be matched using a special `__block_id` label. All other labels are applied to series. This design makes it easy to support multi-tenant environment and match specific blocks.

Author: My thought is to use `tombstone CLI` and has an optional `block id` flag to do equal match. If `block id` is specified multiple times then we make it `__block_id=~(block1|block2|block3)`

> * **label matchers**: label matchers to match the block and series. Like this example, blocks can be matched using external labels and blocks can be matched using a special `__block_id` label. All other labels are applied to series. This design makes it easy to support multi-tenant environment and match specific blocks.

Author: In my experience, sometimes the compactor halts due to some out-of-order chunks from some blocks. Then I can just delete those series using the given block IDs.
Other than that, the use case seems rare.

* **start timestamp**
* **end timestamp**: end timestamp should be the tombstone creation time if not specified, otherwise the compactor needs to wait for a long time to make sure all matched blocks are processed. Considering the tombstone pending period, the end timestamp can be set to `(now + pending period)` as well. We don't consider history blocks backfilling now.
* **author**
* **reason**

> ### Store Gateway Masks Series based on Tombstones

Author: Yes, it was implemented in https://github.com/thanos-io/thanos/pull/3075 but hasn't been finished. AFAIU, we agree on the high level idea of the initial proposal but for the implementation details, since the pr is not merged I won't consider it as implemented.

Like block metadata, the store gateway also fetches tombstones in a configurable interval, which means some delay is expected: user creates a tombstone but she is still able to query the data to be deleted.

There are different ways to mask series based on tombstones. We will discuss two options here:

#### Option 1: Mask Series At Query Time

> During query time, the store gateway decides what blocks to query based on the request min, max time and label matchers. Meanwhile, it can also decide what tombstones are going to apply for the selected blocks based on tombstone time range and label matchers.

Author: Thanks! Updated.

Then during the query time, it checks the series label sets against the tombstone matchers, as well as the chunks and tombstone time range to mask the series data.

Pros:
* Easy to implement and very straightforward
* Less memory usage.

Cons:
* Slower at query time

> #### Option 2: Mask Series At Tombstone Sync Time

Author: You are right. 🤣 Maybe I shouldn't talk about the series masking thing as the two options are almost the same, just the implementation details.

```go
// Result for the processed result for each tombstone, each block.
type MemTombstones struct {
    intvlGroups map[storage.SeriesRef]Intervals
    mtx         sync.RWMutex
}

// Result for the processed result for each block.
// Key: global tombstone ID, Value: processed result for each tombstone.
type BlockTombstones map[ulid.ULID]*MemTombstones
```

Since all the blocks and tombstones are immutable, this result can be easily cached in memory or in filesystem. Ideally for each tombstone, each block we just need to calculate once. Since `BlockTombstones` is a map, it is easy to add or delete one key if users want to add more tombstones or remove one tombstone.

During the query time, the cached tombstone will be used. As it knows which `SeriesRef`, which intervals to apply, it is very straightforward to mask series.

Pros:
* More efficient during query time, less processing required.
* As data can be cached there is not much computation required.

Cons:

> * More memory required.

Reviewer 3: Would it be possible to use the filesystem instead, similar to how store-gw downloads the index file?

Author: Yes, since latency is not really bottleneck for compactor we can definitely store those data in the filesystem. But those files shouldn't be too large as it contains series ids and intervals only. To optimize the space further we can use bitmap or bloom filter.

* If read is much less than write, it is inefficient as the pre-calculated result won't be used much.

#### Considerations

Considering the infrequent change of tombstones, option 2 is preferred as all data can be cached and it is faster during query time. 

> ### Tombstone Deletion with Compactor

Reviewer 4: Reading further that store series masking seems like a good "additonal context", not a proposal. Should we put it in "Additional Context" section then?

> ### Tombstone Deletion with Compactor

Reviewer 3: Implementing removal in compactor is already going providing value, so I agree that masking can be done as an iterative improvement.

> ### Tombstone Deletion with Compactor

Author: Masking was proposed in the previous proposal. I think it would be better to go back and discuss masking there in the old proposal... Instead of talking about it here. I am not sure if it is okay to modify the old one.

Compactor performs the job of deleting the series from object storage and cleans up tombstone files.

The main challenge is about scalability. The reasons are:
* Global tombstone files makes the compactor difficult to know which blocks match the tombstones. It needs to scan all the block indexes to find the matched blocks.
* If one block has too many matched series to mask, the query performance can be hugely impacted. In this case, compactor planner should be able to know the ratio of the masked series / total series. If the ratio is high, the block needs to be compacted and masked series should be removed from the block.

There are other questions about how we should perform the deletion along with compaction:
* Series are deleted at normal compaction. This might need to wait a long time and hard to apply for blocks at the max compaction level.
* Rewrite blocks separately. This needs more coordination with compaction to avoid duplicate/unnecessary work.

> To overcome those challenges, we propose two options.

Reviewer 4: Can we propose one option (pick the one you like the most) and put rest to alternatives?

> To overcome those challenges, we propose two options.

Author: Thanks!

#### Option 1: Compactor Scans Blocks

> Like store gateway, compactor can also scan block indexes to find the blocks matching the tombstones. Similar to the `Mask Series At Tombstone Sync Time` of store gateway, compactor can calculate the processed tombstones for each block, each tombstone and cache the results locally. As blocks and tombstones are immutable, the result needs to be calculated only once and then reused easily. As the Prometheus format tombstones are calculated, during the compaction time the compactor just needs to generate the tombstone file in the filesystem and the TSDB compactor will apply the deletion.

Reviewer 5: If there is a block that has series that needs to be deleted, the block is going to be compacted again? But this time its only one block being sent to the compactor with an additional tombstone file? Am I understanding correctly?

Author: What do you mean by compacted again?
This only happens once and tsdb compaction by default takes care of series deletion if there is any tombstones.

But this only applies to blocks that don't reach compaction max level. For max level block we need to schedule separate compaction for them in the planner.

There are two approaches about how to scan block indexes:

> 1. Download all the blocks locally and check against the tombstones.

Reviewer 4: We cannot download all blocks locally - we might not have enough disk for this. Should we instead scan one by one periodically? 

Would be nice to just think now how often we sync and what's the pseudoalgorithm for this. This would give us some idea about cost and delays - and tell us if we can start with option 1 or we should move to option 2 straight away 🤗

2. Use part of the store gateway code to scan the block index and chunks on the object storage to generate `MemTombstones` result.

Approach 2 should be better as only block indexes need to be downloaded, chunks data are reading from the object storage on the fly so disk space can be saved. As compactor is an offline component, it is acceptable to read from the object storage rather than downloading the whole data.

> The pitfall is that we need to expose part of the store gateway code like `BucketIndexReader`, `BucketChunkReader` as well as other internal code to be used by the compactor.

Reviewer 4: That's fine - it's part of the same repo.

> #### Option 2: New API to Match Tombstones

Reviewer 4: I like it - it feels like low-effort clear API to me 👍🏽

Reviewer 4: I think we could propose it as an optimization if the option 1 will opt 1 will proven too be slow / too expensive in terms of CPU/mem/objstorage access

> Since the block indexes are already downloaded by the store gateway, it can already scan the blocks and calculate the tombstones. So a way here is to expose some APIs from store gateway, and then compactor can query the APIs to know what blocks can be deleted.

Reviewer 4: It's worth to mention that this only works if Store overreach all blocks you want to work with.

Author: As mentioned below, compactors talk to queriers and queriers federate responses from stores. This way should work but a little bit complicated...

The gRPC API is called `Tombstone` API. The compactor can either query from the store gateway directly or query from queriers, and queriers do the API federation from multiple store gateway instances.

The example gRPC format looks like below. The response is almost the same as the `MemTombstone` result.

```protobuf
service Tombstone {
  rpc Tombstone(TombstoneRequest) returns (TombstoneResponse);
}

message TombstoneRequest {
  int64 min_time = 1;
  int64 max_time = 2;

>   repeated LabelMatcher matchers = 3 [(gogoproto.nullable) = false];

Reviewer 4: In some way, compactor is responsible only for some blocks (some streams). Should we perhaps give that info in requests? (maybe I am overcomplicating)?

Author: Yeah you are totally right, I don't have any idea about this so far but if we go with the API direction we definitely need to design this.
Maybe an optional block ID array can be added to only generate tombstones for those blocks if matched? Or we can add block IDs to matchers.

}

message TombstoneResponse {
  repeated SeriesIntervals series_intervals = 1;
}

message SeriesIntervals {
  int64 series_ref = 1;
  repeated Interval intervals = 2;
}

message Interval {
  int64 start = 1;
  int64 end = 2;
}
```

After getting the query result, the compactor does almost the same thing as the previous option: generate the tombstone file during compaction time.

The drawback for this approach is that we are inventing a new API for this purpose only, which might be a bit of pain in the future.

> #### Option 3: Delete Series using Rewrite Logic

Reviewer 4: ```suggestion
#### Option 3: Delete Series using `thanos tools block rewrite` CLI
```

The previous two options do the series deletion at compaction time, this option does actual series deletion using block rewrite. For the index scanning, it is still required to choose either option 1 or 2 to avoid unncessary rewrite even if the block doesn't have any matched series.

> This is not so efficient as rewrite is a different process than compaction. We need to add this step and make sure rewrite can work together with compaction and no redundant work is done.

Reviewer 2: nit:
```suggestion
This is not efficient as rewrite is a different process from compaction. We need to add this step and make sure rewrite can work together with compaction and no redundant work is done.
```

Author: Thanks! Updated.

#### Considerations

Overall, among the 3 options, option 1 is better as it is easier to implement and straightforward. No additional API required or no additional logic change on the compaction side.

### Other Design Considerations

#### Compaction Planner for Tombstone

> If series deletion is done at compaction time, then no much thing to change at the planner as we just need to delete series at compaction time. For the ratio calculation, as the tombstone results are pre-calculated it is easy to know the ratio as well. The threshold can be made configurable too.

Reviewer 2: nit:
```suggestion
If series deletion is done at compaction time, then there's not much to change about the planner, as we just need to delete series at compaction time. For the ratio calculation, as the tombstone results are pre-calculated, it is easy to know the ratio as well. The threshold can be made configurable too.
```

Author: Thanks! Updated.

> The exception is for blocks that at max compaction level, no compaction will happen to it and the tombstones will remain forever if the ratio is lower than the deletion threshold. In this case the planner can check the block compaction level first, if it is max compaction level and the tombstone exists for some period of time, then do compaction for that block.

Reviewer 5: Is the ratio going to be calculated per each tombstone or based on the total of all tombstones that can act on a particular block? Asking because even if one tombstone doesn't reach above the threshold, the sum of multiple smaller tombstones might.

Author: It is for all tombstones. We calculate the number of series to be deleted for all tombstones that match this block.

```go
// Result for the processed result for each tombstone, each block.
type MemTombstones struct {
	intvlGroups map[storage.SeriesRef]Intervals
	mtx         sync.RWMutex
}

// Result for the processed result for each block.
// Key: global tombstone ID, Value: processed result for each tombstone.
type BlockTombstones map[ulid.ULID]*MemTombstones
```

Since all the blocks and tombstones are immutable, this result can be easily cached in memory or in filesystem. Ideally for each tombstone, each block we just need to calculate once. Because `BlockTombstones` is a map, it is easy to add or delete one key if users want to add more tombstones or remove one tombstone.

During the query time, the cached tombstone will be used. As it knows which `SeriesRef`, which intervals to apply, it is very straightforward to mask series.

Pros:
* More efficient during query time, less processing required.
* As data can be cached there is not much computation required.

Cons:
* More memory required.
* If read is much less than write, it is inefficient as the pre-calculated result won't be used much.

> ##### Considerations

Reviewer 4: What this means? We already added `### Proposed Option"`.

Is this section still relevant?
